src/main/fc/fc_hardfaults.c

```diff
@@ -39,7 +39,7 @@
  * cause of the fault.
  * The function ends with a BKPT instruction to force control back into the debugger
  */
-void hard_fault_handler_c(unsigned long *hardfault_args)
+void __attribute__((optimize("-O0"))) hard_fault_handler_c(unsigned long *hardfault_args)
```

Comment:
```
typedet struct {
  uint32_t r0,r1,r2,r3,r12,lr,pc,psr;
} exception_frame;

void hard_fault_handler_c(exception_frame* frm, uint32_t lr)
{

Reply:
And force realization of register by (Check generated code to make sure it works)
```
__asm__ __volatile__("" :: "m" (&_CFSR), "m" (&_HFSR ), ... );

Reply:
And system registers are accessible through `SCB`, much cleaner than hardcoded addresses ...

Reply:
as per comment made by someone else before this is taken straight from https://mcuoneclipse.com/2012/11/24/debugging-hard-faults-on-arm-cortex-m/ I added HF Handler only which was missing. If you dont see a point in using this remove this PR and hard_fault_handler_c(unsigned long *hardfault_args) becuase why is it there in the first place?

Reply:
Having hardfault handler is useful and this PR is working. I would prefer to replace `-O0` with other mechanism. And commented on the code where I think there is space for improvement ...

Reply:
BTW: There is code in https://github.com/betaflight/betaflight/blob/master/src/main/startup/startup_stm32f3_debug_hardfault_handler.S#L133

Reply:
It's F3 only so we would need to add startup files for the rest. ```-O0``` is not needed on HF Handler indeed, but the other function will be optimized out by lto if you have an idea how to deal with it, then I'm happy to hear it.

Reply:
Eh, LTO is easy: `__attribute__((used))` .. I thought that GCC is optimizing variable assignments ...